### ISSUE TITLE
Enhancement: List a directory content

### ISSUE THREAD

**Issue by guillaumerems** [geky/littlefs#2]
I just implemented the littlefs and it would be great to have a function to list a directory with the following result:
- type (file, dir)
- name
- size

**Comment by geky** [comment]
Hi @guillaumerems, glad to hear you were able to get littlefs working.

You can use the `lfs_dir*` functions to list a directory like this:

``` c
int lfs_ls(lfs_t *lfs, const char *path) {
    lfs_dir_t dir;
    int err = lfs_dir_open(lfs, &dir, path);
    if (err) {
        return err;
    }

    struct lfs_info info;
    while (true) {
        int res = lfs_dir_read(lfs, &dir, &info);
        if (res < 0) {
            return res;
        }

        if (res == 0) {
            break;
        }

        switch (info.type) {
            case LFS_TYPE_REG: printf("reg "); break;
            case LFS_TYPE_DIR: printf("dir "); break;
            default:           printf("?   "); break;
        }

        static const char *prefixes[] = {"", "K", "M", "G"};
        for (int i = sizeof(prefixes)/sizeof(prefixes[0])-1; i >= 0; i--) {
            if (info.size >= (1 << 10*i)-1) {
                printf("%*u%sB ", 4-(i != 0), info.size >> 10*i, prefixes[i]);
                break;
            }
        }

        printf("%s\n", info.name);
    }

    err = lfs_dir_close(lfs, &dir);
    if (err) {
        return err;
    }

    return 0;
}

lfs_ls(&lfs, "/");
```

However, there are many different ways to list a directory, so this function probably won't end up in this repo.

This repo's goal is to provide a simple low-level (posix-like) implementation of littlefs that contains all of the core functionality needed for a filesystem. User-friendly bash functions (such as ls, cp, recursive remove) are left up to higher-level APIs.

**Comment by guillaumerems** [comment]
Thanks @geky it works perfectly
I will continue the tests

**Comment by jeot** [comment]
Hello all,
I was looking for a simple arduino/c++ solution to use within the NodeMCU module.
I noticed the solution provided was a little complicated for me and maybe related to modifying the LittleFS library itself (which I'm not sure, I'm just guessing).
So I have come up with a simple function for listing all the files inside a directory (recursively). 
I hope it helps future noobies like myself! :D

```
void listAllFilesInDir(String dir_path)
{
	Dir dir = LittleFS.openDir(dir_path);
	while(dir.next()) {
		if (dir.isFile()) {
			// print file names
			Serial.print("File: ");
			Serial.println(dir_path + dir.fileName());
		}
		if (dir.isDirectory()) {
			// print directory names
			Serial.print("Dir: ");
			Serial.println(dir_path + dir.fileName() + "/");
			// recursive file listing inside new directory
			listAllFilesInDir(dir_path + dir.fileName() + "/");
		}
	}
}
```
Use like this:
```
if (LittleFS.begin()) {
	listAllFilesInDir("/");
}
```
The output should be something like this:
```
File: /asset-manifest.json
File: /favicon.ico        
File: /index.html
File: /manifest.json      
File: /robots.txt
Dir: /static/
Dir: /static/css/
File: /static/css/main.1635e87c.chunk.css
Dir: /static/js/
File: /static/js/2.2c405b20.chunk.js
File: /static/js/3.8cb6885c.chunk.js
File: /static/js/main.435818a8.chunk.js
File: /static/js/runtime-main.00e2c71a.js
```

**Comment by dsabala** [comment]
@geky Is there another way to iterate over an entire directory (for example, to implement an `ls` utility) without calling lfs_dir_read in a loop and reading the same blocks and pages multiple times? That approach is extremely time-consuming. If the filesystem exposed something like lfs_iterate_dir(callback_fn fn), iterating over a whole directory wouldn’t be so expensive. Am I missing something?

**Comment by BenBE** [comment]
The `lfs_dir_read` function internally already continues where it left of at the previous step. Thus a small wrapper like the following:

```c
/* callback returns 0 to continue, negative to stop and propagate an error */
typedef int (*lfs_dir_cb_t)(lfs_t *lfs, const struct lfs_info *info, void *cb_data);

/* iterate directory and call cb for each entry */
int lfs_dir_iterate(lfs_t *lfs, const char *path, lfs_dir_cb_t cb, void *cb_data) {
    lfs_dir_t dir;
    int err = lfs_dir_open(lfs, &dir, path);
    if (err) {
        return err;
    }

    struct lfs_info info;
    while (true) {
        int res = lfs_dir_read(lfs, &dir, &info);
        if (res < 0) {
            lfs_dir_close(lfs, &dir);
            return res;
        }
        if (res == 0) {
            break;
        }

        int cbr = cb(lfs, &info, cb_data);
        if (cbr < 0) {
            lfs_dir_close(lfs, &dir);
            return cbr;
        }
    }

    err = lfs_dir_close(lfs, &dir);
    if (err) {
        return err;
    }

    return 0;
}
```

will mostly already do what you want. That wrapper can be used like this:

```c
/* sample callback that prints type, human-friendly size, and name */
int print_entry_cb(lfs_t *lfs, const struct lfs_info *info, void *cb_data) {
    (void)lfs;

    static const char *prefixes[] = {"", "K", "M", "G"};

    switch (info->type) {
        case LFS_TYPE_REG: printf("reg "); break;
        case LFS_TYPE_DIR: printf("dir "); break;
        default:           printf("?   "); break;
    }

    for (int i = (int)(sizeof(prefixes)/sizeof(prefixes[0])) - 1; i >= 0; i--) {
        uint32_t shift = 10U * (uint32_t)i;
        uint32_t thresh = (shift >= 31) ? 0x80000000U : (1U << shift);
        if (i == 0 || info->size >= thresh) {
            unsigned val = (shift >= 31) ? info->size : (unsigned)(info->size >> shift);
            printf("%*u%sB ", 4 - (i != 0), val, prefixes[i]);
            break;
        }
    }

    printf("%s\n", info->name);
    return 0;
}
```

```c
/* call to iterate root and print entries */
int rc = lfs_dir_iterate(&lfs, "/", print_entry_cb, NULL);
if (rc) {
    /* handle error */
}
```

The bigger issue most likely is with the internal caching and the requirement that some parts of the directory need to be read multiple times (due to CoW and figuring out the most recent state of things).

@geky I think this thin wrapper might be worth adding in the library itself. Or with a separate recursive variant.

**Comment by dsabala** [comment]
@BenBE Thank you for details, I will give it a try, but I already achieved acceptable `ls` performance by lowering `metadata max` setting to single page size = 2KB to better suite my NAND FLASH with 128KB erasable block.

https://github.com/littlefs-project/littlefs/issues/1148

**Comment by BenBE** [comment]
> [@BenBE](https://github.com/BenBE) Thank you for details, I will give it a try, but I already achieved acceptable `ls` performance by lowering `metadata max` setting to single page size = 2KB to better suite my NAND FLASH with 128KB erasable block.
> 
> [#1148](https://github.com/littlefs-project/littlefs/issues/1148)

Figured as much from your other messages on the tracker. Mostly posted the above snippet as an (rough) example for directory listing with a callback, which might be of interest to others too.

I think the most pressing issue in your setup is really (as hinted in my post), how much LittleFS has to re-read the directory when figuring out the next entry to show. The worst case you could have is a large directory, spanning multiple block reads, where a file is created near the beginning, only to be deleted right before the directory content ends.

```
NEW ITEM 0
NEW NAME 0 file_a
NEW ITEM 1
NEW NAME 1 file_b
…
NEW ITEM 999
NEW NAME 999 file_zzz
DROP ITEM 0
DROP ITEM 1
…
DROP ITEM 998
```

(directory only containing `file_zzz`)

A directory with such a layout would essentially have quadratic complexity to list, and if it causes reads for many blocks to be issued, the latency of each read will add up. Not sure if LFSv3 will improve on the handling of directory metadata needed for listing, but unless you can reduce the number of block reads needed to process the data in the directory, this situation will always be slow.

NB: Example is: creating N files to make the directory span many blocks, deleting in create-order without triggering compaction.

**Comment by geky** [comment]
> [@geky](https://github.com/geky) Is there another way to iterate over an entire directory (for example, to implement an `ls` utility) without calling lfs_dir_read in a loop and reading the same blocks and pages multiple times?

No, `lfs_dir_read` is optimal. I don't think a callback API would lead to any optimizations (we can always shove more state in `lfs_dir_t`), but it's an interesting train of thought.

I'll try to go into detail on the performance implications in https://github.com/littlefs-project/littlefs/issues/1148, but will keep this focused on the API to hopefully organize things.

> [@geky](https://github.com/geky) I think this thin wrapper might be worth adding in the library itself. Or with a separate recursive variant.

Callback driven iteration/traversal APIs are an interesting thought (though low priority). But I'm not sure they're actually easier for users due to the clumsiness of function definitions in C. Maybe it's just me, but everytime I use `lfs_fs_traverse`, I find the code hard to read because of how far the traversal logic is from the calling function.

(
You all ever daydream about your perfect C variant? Mine would include GCC's [nested function extension](https://gcc.gnu.org/onlinedocs/gcc/Nested-Functions.html), or maybe a variant that allows local function definitions but NOT closures. C++ has something similar (captureless lambdas) and they are surprisingly handy for code composition without dragging in the complexity of closure state/managment.
)

It's worth noting in littlefs3 (https://github.com/littlefs-project/littlefs/pull/1111), `lfs_fs_traverse` moves away from the callback API and adopts an incremental dir-like iterator (`lfs3_trv_open/read/close`). If you could only choose one API, the `open/read/close` API is the most flexible as it allows for other operations between iteration steps. You can't iterate of two directories in parallel with callbacks, for example. This is especially useful for traversals as it lets you put them in an idle thread.
